### PR TITLE
Laboratory work #8, Zakharova Sofia - 21FPL2

### PR DESCRIPTION
**Pull request by solonarr**

### CODE REVIEW

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]
I do not see any changes related to lab 8

File: lab_8_sft/main.py [hunk 1]
```diff
@@ -32,12 +65,41 @@ def analyze(self) -> dict:
...
+        }
+
+        return ds_properties
 
     @report_time
     def transform(self) -> None:
         """
         Apply preprocessing transformations to the raw dataset.
         """
+        # self._data = (
```
Comment (artyomtugaryov) [inline lab_8_sft/main.py:84]: remove

File: lab_8_sft/service.py [hunk 2]
```diff
@@ -8,6 +8,7 @@
     print('Library "fastapi" not installed. Failed to import.')
     FastAPI = None
 
+
```
Comment (artyomtugaryov) [inline lab_8_sft/service.py:11]: remove

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_8_sft/service.py (hunk 2 above)
Comment (solonarr) [reply lab_8_sft/service.py:11]: but it is a default piece of code

Comment (solonarr) [reply lab_8_sft/service.py:7]: done done done

File: lab_8_sft/main.py [hunk 3]
```diff
@@ -173,6 +253,20 @@ def infer_sample(self, sample: tuple[str, ...]) -> str | None:
         Returns:
             str | None: A prediction
         """
+        # if not self._model:
```
Comment (artyomtugaryov) [inline lab_8_sft/main.py:256]: remove

File: lab_8_sft/main.py [hunk 4]
```diff
@@ -173,6 +253,20 @@ def infer_sample(self, sample: tuple[str, ...]) -> str | None:
...
+        #     return None
```
Comment (artyomtugaryov) [inline lab_8_sft/main.py:257]: .

File: lab_8_sft/main.py [hunk 5]
```diff
@@ -182,6 +276,36 @@ def infer_dataset(self) -> pd.DataFrame:
         Returns:
             pd.DataFrame: Data with predictions
         """
+        # dataloader = DataLoader(self._dataset, batch_size=self._batch_size)
```
Comment (artyomtugaryov) [inline lab_8_sft/main.py:279]: .

File: lab_8_sft/main.py [hunk 6]
```diff
@@ -194,6 +318,31 @@ def _infer_batch(self, sample_batch: Sequence[tuple[str, ...]]) -> list[str]:
         Returns:
             list[str]: model predictions as strings
         """
+        # texts = [sample[0] for sample in sample_batch]
```
Comment (artyomtugaryov) [inline lab_8_sft/main.py:321]: .

File: lab_8_sft/main.py [hunk 7]
```diff
@@ -209,6 +358,9 @@ def __init__(self, data_path: Path, metrics: Iterable[Metrics]) -> None:
             data_path (pathlib.Path): Path to predictions
             metrics (Iterable[Metrics]): List of metrics to check
         """
+        super().__init__(metrics)
+        # self._loaded_metrics = [load(metric.value) for metric in self._metrics]
```
Comment (artyomtugaryov) [inline lab_8_sft/main.py:362]: .

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

**Comment by solonarr** [comment]
image

File: lab_8_sft/main.py [hunk 8]
```diff
@@ -97,6 +154,23 @@ def tokenize_sample(
     Returns:
         dict[str, torch.Tensor]: Tokenized sample
     """
+    tokenized_source = tokenizer(sample[ColumnNames.SOURCE.value],
```
Comment (artyomtugaryov) [inline lab_8_sft/main.py:157]: try to call tokenizer once

File: lab_8_sft/main.py [hunk 9]
```diff
@@ -217,6 +369,24 @@ def run(self) -> dict | None:
...
+            if metric == Metrics.ROUGE.value:
+                metric_evaluator = load(metric.value, seed=77)
+            else:
+                metric_evaluator = load(metric.value)
+            calculated = metric_evaluator.compute(predictions=predictions, references=references)
+
+            if metric.value == Metrics.ROUGE.value:
+                scores[metric.value] = float(calculated['rougeL'])
+            else:
+                scores[metric.value] = float(calculated['bleu'])
```
Comment (artyomtugaryov) [inline lab_8_sft/main.py:387]: use `metric.value`

File: lab_8_sft/main.py [hunk 10]
```diff
@@ -233,8 +403,73 @@ def __init__(self, model_name: str, dataset: Dataset, sft_params: SFTParams) ->
...
+            output_dir=str(self._finetuned_model_path),
+            per_device_train_batch_size=self._batch_size,
+            max_steps=self._max_sft_steps,
+            learning_rate=self._learning_rate,
+            save_strategy="no",
+            use_cpu=bool(self._device == "cpu"),
+            load_best_model_at_end=False
+        )
+
+        # trainer = Trainer(
```
Comment (artyomtugaryov) [inline lab_8_sft/main.py:445]: .

File: lab_8_sft/main.py [hunk 11]
```diff
@@ -233,8 +403,73 @@ def __init__(self, model_name: str, dataset: Dataset, sft_params: SFTParams) ->
...
+        # )
+
+        trainer = Trainer(
+            model=self._model,
+            args=training_args,
+            train_dataset=self._dataset,
+        )
+
+        trainer.train()
+        # var 1
```
Comment (artyomtugaryov) [inline lab_8_sft/main.py:458]: .

File: lab_8_sft/main.py [hunk 12]
```diff
@@ -233,8 +403,73 @@ def __init__(self, model_name: str, dataset: Dataset, sft_params: SFTParams) ->
...
+        # AutoTokenizer.from_pretrained(self._model_name)
+        # .save_pretrained(self._finetuned_model_path)
+
+        # var 3
+        self._model.merge_and_unload()
+        self._model.save_pretrained(self._finetuned_model_path,
+                                    safe_serialization=False)
+        AutoTokenizer.from_pretrained(self._model_name).save_pretrained(self._finetuned_model_path)
+
+        # T5TokenizerFast.from_pretrained(self._model_name)
```
Comment (artyomtugaryov) [inline lab_8_sft/main.py:474]: .

File: lab_8_sft/start.py [hunk 13]
```diff
@@ -1,16 +1,180 @@
+# """
```
Comment (artyomtugaryov) [inline lab_8_sft/start.py:1]: ?

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_8_sft/main.py [hunk 14]
```diff
@@ -233,8 +403,74 @@ def __init__(self, model_name: str, dataset: Dataset, sft_params: SFTParams) ->
...
+        # var 2
+        # self._model.save_pretrained(self._finetuned_model_path)
+        # AutoTokenizer.from_pretrained(self._model_name)
+        # .save_pretrained(self._finetuned_model_path)
+
+        # var 3
+        self._model.merge_and_unload()
+        # self._model.save_pretrained(self._finetuned_model_path,
+        #                             safe_serialization=False)
+        self._model.save_pretrained(self._finetuned_model_path)
```
Comment (artyomtugaryov) [inline lab_8_sft/main.py:472]: `trainer.model.base_model.save_pretrained(finetuned_model_path)`

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_8_sft/main.py (hunk 9 above)
Comment (solonarr) [reply lab_8_sft/main.py:387]: done

File: lab_8_sft/start.py (hunk 13 above)
Comment (solonarr) [reply lab_8_sft/start.py:1]: done

File: lab_8_sft/main.py (hunk 12 above)
Comment (solonarr) [reply lab_8_sft/main.py:474]: done

File: lab_8_sft/main.py (hunk 10 above)
Comment (solonarr) [reply lab_8_sft/main.py:445]: done

File: lab_8_sft/main.py (hunk 7 above)
Comment (solonarr) [reply lab_8_sft/main.py]: done

File: lab_8_sft/main.py (hunk 6 above)
Comment (solonarr) [reply lab_8_sft/main.py]: done

File: lab_8_sft/main.py (hunk 5 above)
Comment (solonarr) [reply lab_8_sft/main.py]: done

File: lab_8_sft/main.py (hunk 4 above)
Comment (solonarr) [reply lab_8_sft/main.py:284]: done

File: lab_8_sft/main.py [hunk 15]
```diff
@@ -97,6 +157,37 @@ def tokenize_sample(
     Returns:
         dict[str, torch.Tensor]: Tokenized sample
     """
+    # tokenized_source = tokenizer(sample[ColumnNames.SOURCE.value],
```
Comment (artyomtugaryov) [inline lab_8_sft/main.py:160]: .

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]
